### PR TITLE
schema term in expected response

### PR DESCRIPTION
**Pull request by sebastiankb**
also see #1053

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/wot-thing-description/pull/1341.html" title="Last updated on Dec 21, 2021, 4:38 PM UTC (00eef3d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-thing-description/1341/21ce84a...00eef3d.html" title="Last updated on Dec 21, 2021, 4:38 PM UTC (00eef3d)">Diff</a>

### CODE REVIEW

**Comment by sebastiankb** [comment]
@benfrancis I double check the issue #1053 . It seems there is a mixup with `ExpectedResponse` and `AdditionalExpectedResponse`.  The examples in #1053 relying on `AdditionalExpectedResponse` where we already have `schema` term in there. It seems that we do not need anymore this PR

**Comment by benfrancis** [comment]
Please see https://github.com/w3c/wot-thing-description/issues/302 and https://github.com/w3c/wot-profile/pull/91 for more detail on use cases

**Comment by egekorkan** [comment]
I think we should solve this in a better way. If we have schema in forms level, for the response of an action, we can have the schema in the output at the action level or in the forms level. Putting more and more "stuff" into forms make the whole WoT more protocol specific. I would like to have maybe query/input and query/output in the action level rather than in the form level. That was part of my hypermedia proposal at https://github.com/w3c/wot-thing-description/tree/main/proposals/hypermedia-control-2

**Comment by benfrancis** [comment]
@egekorkan wrote:
> I think we should solve this in a better way. If we have schema in forms level, for the response of an action, we can have the schema in the output at the action level or in the forms level. Putting more and more "stuff" into forms make the whole WoT more protocol specific. I would like to have maybe query/input and query/output in the action level rather than in the form level. That was part of my hypermedia proposal at https://github.com/w3c/wot-thing-description/tree/main/proposals/hypermedia-control-2

That seems reasonable. Data schemas do already exist at the forms level for `AdditionalExpectedResponse`, but I agree it might be cleaner for a protocol-agnostic schema to be provided at the interaction affordance level.

Does a query action operation really need a separate input and output though? Could there just be one data schema for querying an action status like reading a property?

```
{
  "actions": {
    "fade": {
      "input": {...},
      "output": {...},
      "query": {...}
    }
  }
}
```

It does also beg the question of how a Consumer should interpret the meaning of the `queryaction` response payload, since it could contain anything. In the Core Profile those semantics are defined out of band in the specification.

Note that we don't have the need for a "cancel" data schema in the Core Profile since there's no payload (just a `DELETE` request on the action URL), but I can imagine that might be needed for some protocol bindings.

**Comment by egekorkan** [comment]
@benfrancis wrote: 
> Does a query action operation really need a separate input and output though? Could there just be one data schema for querying an action status like reading a property?

I think that is also possible. We can start with the simple approach and improve if there are limitations.

> It does also beg the question of how a Consumer should interpret the meaning of the queryaction response payload, since it could contain anything.

I would imagine some semantic annotations would be needed if this needs to be flexible.

Note: We can be flexible and prescriptive on anything about hypermedia, there is not a really widely adopted way to do it.

File: index.html [hunk 1]
```diff
@@ -1189,22 +1189,22 @@ <h2>Core Vocabulary Definitions</h2>
                 those defined in <code>securityDefinitions</code>.
                 These must all be satisfied for access to
                 resources.</td><td>mandatory</td><td><a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a> or <a>Array</a> of <a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a></td></tr>
-<tr class="rfc2119-table-assertion" id="td-vocab---Thing"><td><code></code></td><td>Set of named security configurations
+<tr class="rfc2119-table-assertion" id="td-vocab-securityDefinitions--Thing"><td><code>securityDefinitions</code></td><td>Set of named security configurations
                 (definitions only). Not actually applied unless
                 names are used in a <code>security</code>
-                name-value pair.</td><td>mandatory</td><td><a href="#securityscheme"><code>SecurityScheme</code></a> or <a>Array</a> of <a href="#securityscheme"><code>SecurityScheme</code></a></td></tr>
+                name-value pair.</td><td>mandatory</td><td><a>Map</a> of <a href="#securityscheme"><code>SecurityScheme</code></a></td></tr>
```
Comment (mlagally) [inline index.html:1195]: That's a significant change for implementations.
How does this change correspond to the expected response and the issue that is addressed here?
It looks like an unrelated bug fix.

File: index.html [hunk 2]
```diff
@@ -1189,22 +1189,22 @@ <h2>Core Vocabulary Definitions</h2>
...
 <tr class="rfc2119-table-assertion" id="td-vocab-profile--Thing"><td><code>profile</code></td><td>Indicates the WoT Profile mechanisms followed by this Thing Description and the corresponding Thing implementation.</td><td>optional</td><td><a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#anyURI"><code>anyURI</code></a> or <a>Array</a> of <a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#anyURI"><code>anyURI</code></a></td></tr>
-<tr class="rfc2119-table-assertion" id="td-vocab-schemaDefinitions--Thing"><td><code>schemaDefinitions</code></td><td>Set of named data schemas.
-                To be used in a <code>schema</code>
-                name-value pair inside an 
-                <code>AdditionalExpectedResponse</code> object.</td><td>optional</td><td><a>Map</a> of <a href="#dataschema"><code>DataSchema</code></a></td></tr></tbody></table><p><span class="rfc2119-assertion" id=
+<tr class="rfc2119-table-assertion" id="td-vocab-schemaDefinitions--Thing"><td><code>schemaDefinitions</code></td><td>Set of global-based named data schemas.
+                To be used in a local <code>schema</code>
```
Comment (mlagally) [inline index.html:1198]: What is a "global-based" and "local" security schema?
The term global-based is not used anywhere else in the document - It would be good to explain.

File: index.html [hunk 3]
```diff
@@ -2395,7 +2396,13 @@ <h2>Hypermedia Controls Vocabulary Definitions</h2>
...
                 (e.g., <code>text/plain</code>) and potential
                 parameters (e.g., <code>charset=utf-8</code>) for
-                the media type [<cite><a class="bibref" data-link-type="biblio" href="#bib-rfc2046" title="Multipurpose Internet Mail Extensions (MIME) Part Two: Media Types">RFC2046</a></cite>].</td><td>mandatory</td><td><a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a></td></tr></tbody></table></section>
+                the media type [<cite><a class="bibref" data-link-type="biblio" href="#bib-rfc2046" title="Multipurpose Internet Mail Extensions (MIME) Part Two: Media Types">RFC2046</a></cite>].</td><td>mandatory</td><td><a href="http://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/#string"><code>string</code></a></td></tr>
+<tr class="rfc2119-table-assertion" id="td-vocab-schema--ExpectedResponse"><td><code>schema</code></td><td>Used to define the output data schema
```
Comment (mlagally) [inline index.html:2400]: This is the same text as for the "AdditionalExpectedResponse".
What is the difference between ExpectedResponse and AdditionalExpectedResponse?
Why does only the AdditionalExpectedResponse have a status code?
What does "default" mean for these two types?

File: validation/td-validation.ttl [hunk 4]
```diff
@@ -170,26 +170,26 @@
...
+                <code>ExpectedResponse</code> and <code>AdditionalExpectedResponse</code> object."""^^rdf:HTML ;
         sh:node :DataSchemaShape ;
         sh:order 18 ;
     ] ;
     skos:scopeNote """<p><span class="rfc2119-assertion" id=
           "td-context-ns-thing-mandatory">The <code>@context</code>
           name-value pair <em class="rfc2119" title=
           "MUST">MUST</em> contain the anyURI
-          <code>https://www.w3.org/2019/wot/td/v1</code> either
+          <code>https://www.w3.org/2022/wot/td/v1.1</code> either
```
Comment (mlagally) [inline validation/td-validation.ttl:184]: That seems to be an unrelated change, which should be in a separate PR.

File: index.html (hunk 1 above)
Comment (mlagally) [reply index.html:1195]: How is backwards compatibility with TD1.0 guaranteed?

Comment (sebastiankb) [reply index.html:1195]: This is a bug fix which was already fixed in a previous PR. I dont' know why this shows-up again but this change is correct and fulfill TD 1.0:

image

File: index.html (hunk 3 above)
Comment (sebastiankb) [reply index.html:2400]: > This is the same text as for the "AdditionalExpectedResponse".

Yes, it the same intention.

> What is the difference between ExpectedResponse and AdditionalExpectedResponse?

ExpectedResponse can be defined only once. AdditionalExpectedResponse can be defined as array and may be need to define, e.g., different error responses messages (I think this was also requested by you).

>  Why does only the AdditionalExpectedResponse have a status code?

Adding this to ExpectedResponse would violate backwards compatibility.

> What does "default" mean for these two types?

See https://w3c.github.io/wot-thing-description/#sec-default-values

File: validation/td-validation.ttl (hunk 4 above)
Comment (sebastiankb) [reply validation/td-validation.ttl:184]: strange. It seems a race condition with other PRs. The namespace https://www.w3.org/2022/wot/td/v1.1 is correct

**Comment by sebastiankb** [comment]
from today's TD call:
- it seems this PR is not be stable enough to be merged
- it needs further discussion about the motivation 
- decided no to have this for the 2nd WD

**Comment by egekorkan** [comment]
We will close this but it is referenced at https://github.com/w3c/wot/blob/main/planning/ThingDescription/td-next-work-items/usability-and-design.md#reusable-element-design . A new discussion will be started around this whole topic since it is more than just reusable schemas.
